### PR TITLE
Adding DFE class

### PR DESCRIPTION
**Pull request by izabelcavassim**
Adding a DFE class to stdpopsim. The structure is similar to how DemographicModel class was built. 
This is the first attempt, and still not working properly. I need some help to integrate the stuff that is in ext/ to this new class. So we won't necessarily need to do stdpopsim.ext.MutationType(), to specify a given mutation type, for instance.
Maybe @grahamgower can help me with that?

Thanks!

### CODE REVIEW

**Comment by petrelharp** [comment]
I can help!

**Comment by petrelharp** [comment]
>  So we won't necessarily need to do stdpopsim.ext.MutationType(), to specify a given mutation type, for instance.

I'm not sure what you mean - do you mean not having to call `stdpopsim.ext.MutationType( )` [here](https://github.com/popsim-consortium/stdpopsim/pull/1002/files#diff-969ce029f7c31b37201e94e249ff493e10fd349a7124aaa48ff5aadc49da531cR27), for instance? If not, how are you thinking it would work?

File: stdpopsim/species.py [hunk 1]
```diff
@@ -219,6 +229,29 @@ def add_demographic_model(self, model):
             )
         self.demographic_models.append(model)
 
+    def get_dfe(self, id):
+        """
+        Returns a dfe with the specified ``id``.
```
Comment (petrelharp) [inline stdpopsim/species.py:234]: I'd have thought we'd capitalize DFE, so `get_DFE` and `Returns a DFE`? It's an abbreviation?

File: stdpopsim/models.py [hunk 2]
```diff
@@ -299,3 +299,67 @@ def __init__(self, NA, N1, N2, T, M12, M21):
...
+        *,
+        id,
+        description,
+        long_description,
+        mutation_types=None,
+        proportions=None,
+        citations=None,
+        # qc_model=None, # we may want to include qc of dfe in the future
+    ):
+        self.id = id
```
Comment (petrelharp) [inline stdpopsim/models.py:345]: Should we check that `mutation_types` and `proportions` are of the same length, and raise an error if not?

**Comment by petrelharp** [comment]
Looking good so far!

File: stdpopsim/species.py [hunk 3]
```diff
@@ -117,6 +123,9 @@ class Species:
     :ivar demographic_models: This list of :class:`DemographicModel`
         instances in the catalog for this species.
     :vartype demographic_models: list
+    :ivar dfes: This list of :class:`DFE`
+        instances in the catalog for this species.
+    :vartype demographic_models: list
```
Comment (petrelharp) [inline stdpopsim/species.py:128]:
```suggestion
    :vartype dfes: list
```

File: stdpopsim/catalog/HomSap/dfes.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    gamma_shape = 0.186  # shape
+    gamma_mean = -0.01314833  # expected value
+    h = 0.5  # dominance coefficient
+    negative = stdpopsim.ext.MutationType(
+        dominance_coeff=h,
+        distribution_type="g",  # gamma distribution
+        distribution_args=[gamma_mean, gamma_shape],
+    )
+
+    # neutral mutations have 0 proportion because they are not simulated by SLiM
```
Comment (petrelharp) [inline stdpopsim/catalog/HomSap/dfes.py:37]: hm, but it doesn't have 0 proportion below?

File: stdpopsim/models.py (hunk 2 above)
Comment (petrelharp) [reply stdpopsim/models.py:345]: Also check that proportions are sensible (eg not negative, don't sum to more than 1).

Comment (petrelharp) [reply stdpopsim/models.py:345]: and check that all the things in `mutation_types` are of class MutationType

**Comment by petrelharp** [comment]
TODO:
- fixup module import order
- add a flag to the CLI for choosing a DFE (maybe a different PR?)
- add test for DFEs like [this one](https://github.com/popsim-consortium/stdpopsim/blob/main/tests/test_species.py#L119)
- add test that uses a DFE to [test_slim_engine.py](https://github.com/popsim-consortium/stdpopsim/blob/main/tests/test_slim_engine.py)

and write new tests in [test_genomes.py](https://github.com/popsim-consortium/stdpopsim/blob/main/tests/test_genomes.py) -- see [test_models.py](https://github.com/popsim-consortium/stdpopsim/blob/main/tests/test_models.py) for inspiration. Like:
- make sure that the info that we pass in to a DFE object is there in a correct way, e.g., if we specify the `id` then get the `.id` property then its the same; same for mtuation types
- check that it's ok to have proportions that sum to less than one, but can't have negative numbers or sum to more than 1 (maybe we change this?)
- test that we can print it out

**Comment by izabelcavassim** [comment]
TODO (updated my side):

- [ ]  fixup module import order
- [ ]  add a flag to the CLI for choosing a DFE (maybe a different PR?)
- [ ] add test for DFEs like this one
- [ ] add test that uses a DFE to test_slim_engine.py (OR NOT since this will be done in test_genomes.py)
- [ ] add a DFE argument to contig.add_genomic_element_type that pulls the proportions and mutation types

and write new tests in test_genomes.py -- see test_models.py for inspiration. Like:

- [ ] make sure that the info that we pass in to a DFE object is there in a correct way, e.g., if we specify the id then get the .id property then its the same; same for mtuation types
- [ ] check that proprtions must sum to 1 and not be negative
- [ ] test that we can print it out
- [ ] automatic test that the DFE runs in a simulation as in test_models.py

**Comment by petrelharp** [comment]
>  add a DFE argument to contig.add_genomic_element_type that pulls the proportions and mutation types

Do we still want to do this? I think `contig.add_DFE` does everything we want here?

File: stdpopsim/slim_engine.py [hunk 5]
```diff
@@ -1132,7 +1132,7 @@ def simulate(
         if slim_burn_in < 0:
             raise ValueError("slim_burn_in must be non-negative")
 
-        if slim_scaling_factor != 1:
+        if slim_scaling_factor > 1:
```
Comment (petrelharp) [inline stdpopsim/slim_engine.py:1135]: Hm, I'm not sure about this change - what prompted it?

File: tests/test_genomes.py [hunk 6]
```diff
@@ -72,3 +73,222 @@ def test_add_genomic_element_type(self):
...
+    """
+
+    def test_id_valid(self):
+        assert utils.is_valid_dfe_model_id(self.dfe.id)
+
+
+qc_test_classes = []
+for species in stdpopsim.all_species():
+    for model in species.demographic_models:
+        for dfe in species.dfes:
```
Comment (petrelharp) [inline tests/test_genomes.py:281]: ah, I see! Nice work figuring this out. But, I don't think we need to test all combinations of DFE and model. I don't think we need to even test DFEs with non-generic models at all - they're pretty much completely separated?

Comment (petrelharp) [reply tests/test_genomes.py:281]: Hm - and, this won't test DFEs at all in species without demographic models. I'm replacing this with a generic model.

**Comment by petrelharp** [comment]
I may have gotten the tests working - there seem to be some other slim-related errors for certain demographic models, but that's separate from this PR.

**Comment by izabelcavassim** [comment]
Fixed the coverage and rebased. 
it is all yours @petrelharp, you can read it through!

Thank you 🐢🐢

**Comment by petrelharp** [comment]
Looks good (still)! I tidied up a few more loose ends, and will squash again.

**Pull request merged**